### PR TITLE
Windows debugger: Added dce command. Fixed dmm commands.

### PR DESCRIPTION
**Pull request by zutle**
This PR is to pay for the pizza at r2con. Thanks pancake :)

Added new command on Windows:
`dce`: Continue execution (pass exception to program)

Fixed the following commands on Windows:
`.dmm*`
`dmm` (now reports sane paths)
`dmmj` (when used from e.g. r2pipe)

The `dce` command will instruct the debugged process to handle the raised exception (similar to `shift+F9` in OllyDbg). This is especially useful when debugging protected executables and malware that make use of anti-debugging techniques.

`dmm*` did not generate valid commands on Windows. I have found it extremely useful to quickly get the exports of all loaded DLLs by using `.dmm*`.

The `dmm` command worked but its output would include the name of the executable twice.

The `dmmj` command was problematic when scripting r2 from e.g. Python via r2pipe, because the backspaces in the paths were not properly escaped.

### CODE REVIEW

File: libr/core/cmd_debug.c [hunk 1]
```diff
@@ -806,8 +812,18 @@ static void cmd_debug_modules(RCore *core, int mode) { // "dmm"
 			}
 			break;
 		case 'j':
+#if __WINDOWS__
```
Comment (XVilka) [inline libr/core/cmd_debug.c]: Please add an exception here for __CYGWIN__

File: libr/core/cmd_debug.c [hunk 2]
```diff
@@ -792,8 +792,14 @@ static void cmd_debug_modules(RCore *core, int mode) { // "dmm"
 				//r_cons_printf ("fs+module_%s\n", fn);
 				r_cons_printf ("f mod.%s = 0x%08"PFMT64x"\n",
 						fn, map->addr);
+#if __WINDOWS__
```
Comment (radare) [inline libr/core/cmd_debug.c]: we should escape the map->file path here, i know this is out of the scope of the PR, just something TODO (if you mind in this PR)

Comment (radare) [inline libr/core/cmd_debug.c]: use r_str_escape() and use single quotes

File: libr/core/cmd_debug.c [hunk 3]
```diff
@@ -806,8 +812,18 @@ static void cmd_debug_modules(RCore *core, int mode) { // "dmm"
...
+			{
+				/* Single backslashes cause issues when parsing JSON output, so escape them */
+				char *escaped_path = r_str_escape (map->file);
+				r_cons_printf ("{\"address\":%"PFMT64d",\"name\":\"%s\",\"file\":\"%s\"}%s",
+						map->addr, map->name, escaped_path, iter->n?",":"");
+				free (escaped_path);
+			}
+#else
 			r_cons_printf ("{\"address\":%"PFMT64d",\"name\":\"%s\",\"file\":\"%s\"}%s",
 					map->addr, map->name, map->file, iter->n?",":"");
```
Comment (radare) [inline libr/core/cmd_debug.c]: name is not escaped here. so it may be problematic

File: libr/core/cmd_debug.c [hunk 4]
```diff
@@ -816,8 +832,14 @@ static void cmd_debug_modules(RCore *core, int mode) { // "dmm"
 				//r_cons_printf ("fs+module_%s\n", fn);
 				r_cons_printf ("f mod.%s = 0x%08"PFMT64x"\n",
 						fn, map->addr);
+#if __WINDOWS__
+				/* Use double quotes around the file path on Windows to generate valid commands */
+				r_cons_printf (".!rabin2 -rsB 0x%08"PFMT64x" \"%s\"\n",
+						map->addr, map->file);
```
Comment (radare) [inline libr/core/cmd_debug.c]: same here

File: libr/debug/p/debug_native.c [hunk 5]
```diff
@@ -229,7 +229,14 @@ static void r_debug_native_stop(RDebug *dbg) {
 /* TODO: must return true/false */
 static int r_debug_native_continue(RDebug *dbg, int pid, int tid, int sig) {
 #if __WINDOWS__ && !__CYGWIN__
-	if (ContinueDebugEvent (pid, tid, DBG_CONTINUE) == 0) {
+	/* Honor the Windows-specific signal that instructs threads to process exceptions */
+	DWORD continue_status;
+	if (sig == DBG_EXCEPTION_NOT_HANDLED ) {
```
Comment (radare) [inline libr/debug/p/debug_native.c]: unnecessary space before )

File: libr/debug/p/debug_native.c [hunk 6]
```diff
@@ -229,7 +229,14 @@ static void r_debug_native_stop(RDebug *dbg) {
 /* TODO: must return true/false */
 static int r_debug_native_continue(RDebug *dbg, int pid, int tid, int sig) {
 #if __WINDOWS__ && !__CYGWIN__
-	if (ContinueDebugEvent (pid, tid, DBG_CONTINUE) == 0) {
+	/* Honor the Windows-specific signal that instructs threads to process exceptions */
+	DWORD continue_status;
```
Comment (radare) [inline libr/debug/p/debug_native.c]: use ternary operator here:
```
DWORD continue_status = (sig == DBG_EXCEPTION_NOT_HANDLED)
  ? DBG_EXCEPTION_NOT_HANDLED : DBG_CONTINUE;
```

**Comment by radare** [comment]
The PR looks good so far, but it will be great if you could fix those escaped string issues.

Thanks!

**Comment by zutle** [comment]
Thanks for reviewing my PR so quickly!

I have a question related to memory maps.

You have this struct:
```
typedef struct r_debug_map_t {
	char *name;
	ut64 addr;
	ut64 addr_end;
	ut64 size;
	char *file;
	int perm;
	int user;
} RDebugMap;
```

It has a `name` and a `file`, but these members (at least for the most part) both seem to hold the same string?

So, for Linux you will have this:
```
[0x7f0426ad5309]> dmmj~{}
[
	{
		"address": 4194304,
		"name": "/bin/ls",
		"file": "/bin/ls"
	},
	{
		"address": 139655795195904,
		"name": "/usr/lib/locale/locale-archive",
		"file": "/usr/lib/locale/locale-archive"
	},
	{
		"address": 139655798116352,
		"name": "/lib/x86_64-linux-gnu/libattr.so.1.1.0",
		"file": "/lib/x86_64-linux-gnu/libattr.so.1.1.0"
	},
```

As `name` does not really seem to have a purpose, I have suggested the following in this PR for Windows:
```
[0x77a504f7]> dmmj~{}
[
        {
                "address": 5111808,
                "name": "calc.exe",
                "file": "c:\\Windows\\system32\\calc.exe"
        },
        {
                "address": 2006646784,
                "name": "ntdll.dll",
                "file": "C:\\Windows\\SYSTEM32\\ntdll.dll"
        },
        {
                "address": 2001010688,
                "name": "kernel32.dll",
                "file": "C:\\Windows\\system32\\kernel32.dll"
        },
        {
                "address": 1977090048,
                "name": "KERNELBASE.dll",
                "file": "C:\\Windows\\system32\\KERNELBASE.dll"
        },
```

Also, I have suggested to use the map names as flags, so you will have this on Windows:

```
[0x77a504f7]> dmm*
f mod.calc.exe = 0x000b0000
.!rabin2 -rsB 0x000b0000 "c:\\Windows\\System32\\calc.exe"
f mod.ntdll.dll = 0x779b0000
.!rabin2 -rsB 0x779b0000 "C:\\Windows\\SYSTEM32\\ntdll.dll"
f mod.kernel32.dll = 0x77450000
.!rabin2 -rsB 0x77450000 "C:\\Windows\\system32\\kernel32.dll"
f mod.KERNELBASE.dll = 0x75d80000
.!rabin2 -rsB 0x75d80000 "C:\\Windows\\system32\\KERNELBASE.dll"
f mod.SHELL32.dll = 0x76430000
.!rabin2 -rsB 0x76430000 "C:\\Windows\\system32\\SHELL32.dll"
f mod.msvcrt.dll = 0x75e90000
.!rabin2 -rsB 0x75e90000 "C:\\Windows\\system32\\msvcrt.dll"
f mod.SHLWAPI.dll = 0x77950000
.!rabin2 -rsB 0x77950000 "C:\\Windows\\system32\\SHLWAPI.dll"
f mod.GDI32.dll = 0x76260000
.!rabin2 -rsB 0x76260000 "C:\\Windows\\system32\\GDI32.dll"
```

Which means you will then get the following flag names for the loaded modules:
```
[0x77a504f7]> f~mod.
0x779b0000 0 mod.ntdll.dll
0x75d80000 0 mod.KERNELBASE.dll
0x76430000 0 mod.SHELL32.dll
0x75e90000 0 mod.msvcrt.dll
0x76260000 0 mod.GDI32.dll
0x77590000 0 mod.USER32.dll
0x77220000 0 mod.LPK.dll
0x77080000 0 mod.USP10.dll
0x74740000 0 mod.gdiplus.dll
0x77230000 0 mod.ole32.dll
0x76180000 0 mod.RPCRT4.dll
0x77b40000 0 mod.ADVAPI32.dll
0x77430000 0 mod.sechost.dll
0x77390000 0 mod.OLEAUT32.dll
0x748d0000 0 mod.UxTheme.dll
0x74a30000 0 mod.COMCTL32.dll
0x71210000 0 mod.WINMM.dll
0x75000000 0 mod.VERSION.dll
```

Instead of (old):
```
[0x779e40f1]> f~mod.
0x005d0000 0 mod.c:__Windows__System32__calc.exe
0x779b0000 0 mod.C:__Windows__SYSTEM32__ntdll.dll
0x75d80000 0 mod.C:__Windows__system32__KERNELBASE.dll
0x76430000 0 mod.C:__Windows__system32__SHELL32.dll
0x75e90000 0 mod.C:__Windows__system32__msvcrt.dll
0x76260000 0 mod.C:__Windows__system32__GDI32.dll
0x77590000 0 mod.C:__Windows__system32__USER32.dll
0x77220000 0 mod.C:__Windows__system32__LPK.dll
0x77080000 0 mod.C:__Windows__system32__USP10.dll
0x74740000 0 mod.C:__Windows__WinSxS__x86_microsoft.windows.gdiplus_6595b64144ccf1df_1.1.7601.17514_none_72d18a4386696c80__gdiplus.dll
0x77230000 0 mod.C:__Windows__system32__ole32.dll
0x76180000 0 mod.C:__Windows__system32__RPCRT4.dll
0x77b40000 0 mod.C:__Windows__system32__ADVAPI32.dll
0x77430000 0 mod.C:__Windows__SYSTEM32__sechost.dll
0x77390000 0 mod.C:__Windows__system32__OLEAUT32.dll
0x748d0000 0 mod.c:__Windows__System32__UxTheme.dll
0x74a30000 0 mod.C:__Windows__WinSxS__x86_microsoft.windows.common_controls_6595b64144ccf1df_6.0.7601.17514_none_41e6975e2bd6f2b2__COMCTL32.dll
0x71210000 0 mod.c:__Windows__System32__WINMM.dll
0x75000000 0 mod.c:__Windows__System32__VERSION.dll
```

I like it better this way because my brain can instantly spot which module I'm dealing with instead of parsing those long path name strings :)

What do you think?

**Comment by radare** [comment]
Thanks! merged

**Comment by zutle** [comment]
Awesome! Thanks :)

**Pull request merged**